diff --git a/package.json b/package.json
index e39a2b6..c6b3ed2 100644
--- a/package.json
+++ b/package.json
@@ -44,6 +44,7 @@
   "dependencies": {
     "@ericcornelissen/bash-parser": "^0.5.2",
     "@npmcli/map-workspaces": "^3.0.4",
+    "@pkgjs/parseargs": "github:pkgjs/parseargs",
     "@pnpm/logger": "5.0.0",
     "@pnpm/workspace.pkgs-graph": "2.0.6",
     "@snyk/github-codeowners": "^1.1.0",
@@ -74,6 +75,7 @@
     "@types/minimist": "1.2.3",
     "@types/node": "20.8.3",
     "@types/npmcli__map-workspaces": "3.0.2",
+    "@types/pkgjs__parseargs": "0.10.1",
     "@types/webpack": "5.28.3",
     "@typescript-eslint/eslint-plugin": "6.7.4",
     "@typescript-eslint/parser": "6.7.4",
diff --git a/scripts/create-new-plugin.ts b/scripts/create-new-plugin.ts
index d6c8cb6..cfc5a04 100644
--- a/scripts/create-new-plugin.ts
+++ b/scripts/create-new-plugin.ts
@@ -1,6 +1,6 @@
 import fs from 'node:fs/promises';
 import path from 'node:path';
-import { parseArgs } from 'node:util';
+import { parseArgs } from '@pkgjs/parseargs';
 
 const {
   values: { name },
diff --git a/src/util/cli-arguments.ts b/src/util/cli-arguments.ts
index 48be9e8..77c1b01 100644
--- a/src/util/cli-arguments.ts
+++ b/src/util/cli-arguments.ts
@@ -1,4 +1,4 @@
-import { parseArgs } from 'node:util';
+import { parseArgs } from '@pkgjs/parseargs';
 
 export const helpText = `✂️  Find unused files, dependencies and exports in your JavaScript and TypeScript projects
 
