diff --git a/knip.jsonc b/knip.jsonc
index f471ada9b2..83844be8aa 100644
--- a/knip.jsonc
+++ b/knip.jsonc
@@ -32,6 +32,7 @@
     "ignoreExportsUsedInFile": {
         "enum": true,
         "interface": true,
+        "namespace": true,
         "type": true
     },
     "mocha": false
diff --git a/src/compiler/sys.ts b/src/compiler/sys.ts
index c8f1766030..64e77c320e 100644
--- a/src/compiler/sys.ts
+++ b/src/compiler/sys.ts
@@ -563,7 +563,7 @@ export const ignoredPaths: readonly string[] = ["/node_modules/.", "/.git", "/.#
 
 let curSysLog: (s: string) => void = noop;
 
-/** @internal */
+/** @internal @knipignore */
 export function sysLog(s: string): void {
     return curSysLog(s);
 }
