// vue/compiler-sfc/index.js
const scriptRe = /<script(\b[^>]*)>([\s\S]*?)<\/script>/g;
const templateRe = /<template(\b[^>]*)>([\s\S]*?)<\/template>/g;

function parse({ source = '', filename = '', sourceMap }) {
  if (sourceMap !== false) throw new Error('Expected source maps to be disabled');
  let script = null;
  let scriptSetup = null;
  let template = null;
  let match;

  while ((match = scriptRe.exec(source))) {
    const attrs = match[1];
    const content = match[2];
    if (attrs.includes('setup')) scriptSetup = { content };
    else script = { content };
  }
  scriptRe.lastIndex = 0;

  match = templateRe.exec(source);
  if (match) template = { type: 'template', content: match[2], attrs: {}, start: match.index, end: templateRe.lastIndex };
  templateRe.lastIndex = 0;

  return {
    source,
    filename,
    template,
    script,
    scriptSetup,
    styles: [],
    customBlocks: [],
    cssVars: [],
    errors: [],
  };
}

function parseComponent(source, options = {}) {
  return parse({ source, filename: options.filename, sourceMap: false });
}

function compileTemplate({ source, filename, sourceMap, isProduction, prettify }) {
  if (!filename || sourceMap !== false || !isProduction || prettify !== false) {
    throw new Error('Unexpected Vue 2 template compiler options');
  }
  const branch = { type: 1, tag: 'span', children: [] };
  const component = { type: 1, tag: 'AppleCard', children: [] };
  branch.ifConditions = [{ block: branch }, { block: component }];
  const expression = source.includes('formatApple') ? '_vm.formatApple(_vm.count)' : 'null';
  return {
    ast: { type: 1, tag: 'main', children: [branch] },
    code: `var render = function () { return [${expression}, _vm.$set(_vm.form, 'name', 'apple')] }`,
    source,
    tips: [],
    errors: [],
  };
}

module.exports = { compileTemplate, parse, parseComponent };
